File: core/crates/db/src/repo/match_queue.rs
//! MatchQueue repository implementation

use async_trait::async_trait;
use chrono::{DateTime, Utc};
use sea_orm::*;

use crate::entity::match_queue::{self, Entity as MatchQueue, QueueStatus};
use crate::traits::MatchQueueRepository;
use crate::{Error, Result};

/// SeaORM-based match queue repository
pub struct SeaOrmMatchQueueRepo {
    db: DatabaseConnection,
}

impl SeaOrmMatchQueueRepo {
    pub fn new(db: DatabaseConnection) -> Self {
        Self { db }
    }
}

#[async_trait]
impl MatchQueueRepository for SeaOrmMatchQueueRepo {
    async fn enqueue(&self, request_id: &str, priority: i32) -> Result<match_queue::Model> {
        let model = match_queue::ActiveModel {
            id: Set(uuid::Uuid::new_v4()),
            request_id: Set(request_id.to_string()),
            priority: Set(priority),
            status: Set(QueueStatus::Pending),
            attempts: Set(0),
            last_error: Set(None),
            next_attempt_at: Set(Utc::now()),
            created_at: Set(Utc::now()),
            updated_at: Set(Utc::now()),
        };
        model.insert(&self.db).await.map_err(Error::from)
    }

    async fn fetch_batch(&self, limit: i64) -> Result<Vec<match_queue::Model>> {
        // Fetch pending items that are ready for processing
        let items = MatchQueue::find()
            .filter(match_queue::Column::Status.eq(QueueStatus::Pending))
            .filter(match_queue::Column::NextAttemptAt.lte(Utc::now()))
            .order_by_desc(match_queue::Column::Priority)
            .order_by_asc(match_queue::Column::CreatedAt)
            .limit(limit as u64)
            .all(&self.db)
            .await?;

        // Mark them as processing
        for item in &items {
            let mut active: match_queue::ActiveModel = item.clone().into();
            active.status = Set(QueueStatus::Processing);
            active.updated_at = Set(Utc::now());
            active.update(&self.db).await?;
        }

        Ok(items)
    }

    async fn complete(&self, id: &uuid::Uuid) -> Result<()> {
        let item = MatchQueue::find_by_id(*id)
            .one(&self.db)
            .await?
            .ok_or_else(|| Error::NotFound(format!("Queue item not found: {}", id)))?;

        let mut active: match_queue::ActiveModel = item.into();
        active.status = Set(QueueStatus::Completed);
        active.updated_at = Set(Utc::now());
        active.update(&self.db).await?;
        Ok(())
    }

    async fn fail(&self, id: &uuid::Uuid, error: &str) -> Result<()> {
        let item = MatchQueue::find_by_id(*id)
            .one(&self.db)
            .await?
            .ok_or_else(|| Error::NotFound(format!("Queue item not found: {}", id)))?;

        let retry_delay_secs = 60i64; // Default retry delay
        let mut active: match_queue::ActiveModel = item.clone().into();
        active.attempts = Set(item.attempts + 1);
        active.last_error = Set(Some(error.to_string()));
        active.next_attempt_at = Set(Utc::now() + chrono::Duration::seconds(retry_delay_secs));
        active.status = Set(if item.attempts >= 3 {
            QueueStatus::Failed
        } else {
            QueueStatus::Pending
        });
        active.updated_at = Set(Utc::now());
        active.update(&self.db).await?;
        Ok(())
    }

    async fn count_pending(&self) -> Result<i64> {
        MatchQueue::find()
            .filter(match_queue::Column::Status.eq(QueueStatus::Pending))
            .count(&self.db)
            .await
            .map(|c| c as i64)
            .map_err(Error::from)
    }

    async fn delete_before(&self, cutoff: &DateTime<Utc>) -> Result<u64> {
        let result = MatchQueue::delete_many()
            .filter(match_queue::Column::Status.eq(QueueStatus::Completed))
            .filter(match_queue::Column::UpdatedAt.lt(*cutoff))
            .exec(&self.db)
            .await?;
        Ok(result.rows_affected)
    }
}

File: convex/matchQueue.ts
import { query, mutation } from "./_generated/server";
import { v } from "convex/values";

// List pending items in match queue
export const listPending = query({
  args: { limit: v.optional(v.number()) },
  handler: async (ctx, args) => {
    return await ctx.db
      .query("matchQueueItems")
      .withIndex("by_status", (q) => q.eq("status", "PENDING"))
      .order("desc")
      .take(args.limit ?? 100);
  },
});

// Add to match queue (matches entity/match_queue.rs)
export const add = mutation({
  args: {
    requestId: v.string(),
    priority: v.number(),
  },
  handler: async (ctx, args) => {
    // Check for existing entry
    const existing = await ctx.db
      .query("matchQueueItems")
      .withIndex("by_request", (q) => q.eq("requestId", args.requestId))
      .filter((q) => q.eq(q.field("status"), "PENDING"))
      .first();

    if (existing) {
      return existing._id;
    }

    const now = Date.now();
    return await ctx.db.insert("matchQueueItems", {
      requestId: args.requestId,
      status: "PENDING",
      priority: args.priority,
      attempts: 0,
      lastError: undefined,
      nextAttemptAt: now,
      createdAt: now,
      updatedAt: now,
    });
  },
});

// Mark as processed (completed successfully)
export const markProcessed = mutation({
  args: { id: v.id("matchQueueItems") },
  handler: async (ctx, args) => {
    await ctx.db.patch(args.id, {
      status: "COMPLETED",
      updatedAt: Date.now(),
    });
  },
});

// Mark as failed
export const fail = mutation({
  args: {
    id: v.id("matchQueueItems"),
    error: v.string(),
  },
  handler: async (ctx, args) => {
    const item = await ctx.db.get(args.id);
    if (!item) throw new Error("Queue item not found");

    const now = Date.now();
    const nextAttempt = now + 60000 * Math.pow(2, item.attempts); // Exponential backoff

    await ctx.db.patch(args.id, {
      status: item.attempts >= 3 ? "FAILED" : "PENDING",
      attempts: item.attempts + 1,
      lastError: args.error,
      nextAttemptAt: nextAttempt,
      updatedAt: now,
    });
  },
});

// Get queue stats
export const stats = query({
  args: {},
  handler: async (ctx) => {
    const pending = await ctx.db
      .query("matchQueueItems")
      .withIndex("by_status", (q) => q.eq("status", "PENDING"))
      .collect();

    const processing = await ctx.db
      .query("matchQueueItems")
      .withIndex("by_status", (q) => q.eq("status", "PROCESSING"))
      .collect();

    const completed = await ctx.db
      .query("matchQueueItems")
      .withIndex("by_status", (q) => q.eq("status", "COMPLETED"))
      .collect();

    const failed = await ctx.db
      .query("matchQueueItems")
      .withIndex("by_status", (q) => q.eq("status", "FAILED"))
      .collect();

    return {
      pending: pending.length,
      processing: processing.length,
      completed: completed.length,
      failed: failed.length,
    };
  },
});

// Delete match queue items created before cutoff (matches SeaORM delete_before)
export const deleteBefore = mutation({
  args: { cutoff: v.number() },
  handler: async (ctx, args) => {
    const toDelete = await ctx.db
      .query("matchQueueItems")
      .withIndex("by_created", (q) => q.lt("createdAt", args.cutoff))
      .collect();

    for (const doc of toDelete) {
      await ctx.db.delete(doc._id);
    }
    return toDelete.length;
  },
});
